microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MicroProfileVendorMetricRegistry.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2020, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.wildfly.extension.microprofile.metrics;
+
+import static org.eclipse.microprofile.metrics.MetricRegistry.Type.BASE;
+import static org.eclipse.microprofile.metrics.MetricRegistry.Type.VENDOR;
+import static org.wildfly.extension.metrics.MetricMetadata.Type.COUNTER;
+
+import java.util.Optional;
+import java.util.OptionalDouble;
+
+import io.smallrye.metrics.ExtendedMetadata;
+import io.smallrye.metrics.MetricRegistries;
+import org.eclipse.microprofile.metrics.Counter;
+import org.eclipse.microprofile.metrics.Gauge;
+import org.eclipse.microprofile.metrics.Metadata;
+import org.eclipse.microprofile.metrics.Metric;
+import org.eclipse.microprofile.metrics.MetricType;
+import org.eclipse.microprofile.metrics.MetricUnits;
+import org.eclipse.microprofile.metrics.Tag;
+import org.jboss.as.controller.client.helpers.MeasurementUnit;
+import org.wildfly.extension.metrics.MetricID;
+import org.wildfly.extension.metrics.MetricMetadata;
+import org.wildfly.extension.metrics.MetricRegistry;
+import org.wildfly.extension.metrics.WildFlyMetricMetadata;
+
+public class MicroProfileVendorMetricRegistry implements MetricRegistry {
+
+    final org.eclipse.microprofile.metrics.MetricRegistry vendorRegistry = MetricRegistries.get(VENDOR);
+
+    @Override
+    public void registerMetric(org.wildfly.extension.metrics.Metric metric, MetricMetadata metadata) {
+        final Metric mpMetric;
+        if (metadata.getType() == COUNTER) {
+            mpMetric = new Counter() {
+                @Override
+                public void inc() {
+                }
+
+                @Override
+                public void inc(long n) {
+                }
+
+                @Override
+                public long getCount() {
+                    OptionalDouble value = metric.getValue();
+                    return  Double.valueOf(value.orElse(0)).longValue();
```

Comment:
It would be nice if SmallRye had something like a

```
public interface OptionalMetric extends Metric {

  Optional<Metric> getOptional();
}
```

Then check for that when exporting and if that's the Metric type use the Optional, skipping export !optional.isPresent().

Responding with zero exposes the metric names which in theory could be privileged information.

Reply:
Yes, that's a big difference between the base `metrics` and the `microprofile-metrics-smallrye`subsystem.

In the base metrics, if the metric value is not defined, I skip exposing any info about that (including its name and description): https://github.com/wildfly/wildfly/blob/9c2367ec9fad676658cea070bf6e9e8af95270cf/metrics/src/main/java/org/wildfly/extension/metrics/PrometheusExporter.java#L48

With MicroProfile Metrics, any registered metric is considered valid and I have no point where I can "validated" the metric value before it is printed out.

Reply:
@jmesnil If you think the OptionalMetric idea is useful I can file an RFE for such an SPI in smallrye. It's kind of hacky because it extends 'Metric' just to be an acceptable type for the registry, but the normal Metric impl really isn't meant to be used.